Start linting the test

MVrachev opened the pull request:
Fixes #1582 

**Description of the changes being introduced by the pull request**:

Start linting the files that test the new code. In order to do that the old test files
had to be excluded. They were are renamed to `test_*_old.py`, so we can add globs/regex to exclude
the files testing the old code.
These options will be removed after we remove those files.

Additional `black`, `pylint` and `mypy` errors were found and had to be addressed
before we can start linting.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Rename old test files by adding old suffix

Rename test files testing the old code by adding an "old" suffix.
This is done, so we can easily exclude them from linting.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Fix more black warnings on test files

The changes are automatic linting fixes from black.
The target files are only those who test the new code.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Address more pylint warnings on tests files

All of the changes are made manual.
The target files are only those who test the new code.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Address more mypy warnings on tests files

All of the changes are made manual.
The target files are only those who test the new code.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Start linting test files testing the new code

Exclude regexs/globs are needed to exclude the test files testing
the old code.
After we remove those files we will be able to remove the exclude
regex/globs.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Missing mypy stubs in for dateutil.relativedelta

Mypy reported a couple of errors for missing stubs on GitHub
https://github.com/theupdateframework/python-tuf/runs/4431787287?check_suite_focus=true

The error noted that:
"tests/test_api.py:19: error: Library stubs not installed for
"dateutil.relativedelta" (or incompatible with Python 3.10)"
In order to resolve it I added "types-python-dateutil" into
requirements-test.txt

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

lukpueh commented on tests/test_indefinite_freeze_attack_old.py:
```diff
@@ -36,7 +36,6 @@
   metadata without the client being aware.
 """
 
-import datetime
```
Why this change in an "_old.py" file?

MVrachev replied:
I just saw it was unused and decided to remove it, but I can return it if it looks strange.

lukpueh replied:
🤷  Probably not worth the effort. :) Let's leave it there.

lukpueh commented on tests/test_updater_consistent_snapshot.py:
```diff
@@ -117,6 +121,7 @@ def test_top_level_roles_update(self, test_case_data: Dict[str, Any]):
         sim = self._init_repo(consistent_snapshot)
         updater = self._init_updater(sim)
 
+        # pylint: disable=protected-access
```
Maybe we want to disable this globally for all tests?

MVrachev replied:
I don't think that is possible as of today.
See issue https://github.com/PyCQA/pylint/issues/618 and their project where they keep track of the progress on that issue https://github.com/PyCQA/pylint/projects/1

MVrachev replied:
I don't think that is possible as of today.
See issue https://github.com/PyCQA/pylint/issues/618 and their project where they keep track of the progress on that issue https://github.com/PyCQA/pylint/projects/1

lukpueh replied:
You could call pylint separately for tests and for everything else in tox.ini, as we [do for in-toto](https://github.com/in-toto/in-toto/blob/cb562165b2f62d1fabc80b0b5fdcf212df8fd3d6/tox.ini#L29-L32).

MVrachev replied:
The only downside is that we will no longer have one source for the folders we lint or in other words we won't be able to use `lint_dirs`: https://github.com/theupdateframework/python-tuf/blob/aebd8b0fafdd52e5d6765f4dd628df0492179c1a/tox.ini#L41 for pylint.
@jku any thoughts here as you suggested `lint_dirs`?

jku replied:
globally disabling the lint is ok ... but at this point might be overkill: As an example, there's already a PR that removes these protected accesses (because there is a better design for these cases): I think we won't have more than a couple in the whole test suite next week.

(this is a response to earlier comment)

lukpueh reviewed (approved):
Thanks for the PR, @MVrachev! Looks like a straight-forward set of changes. I left some minor suggestions, but will approve regardless.

I see you adopted the filename changes (add "_old" infix) in some doc headers but not all. I suggest to do this consistently and I'm fine with not updating the doc headers, given that these files are bound to be removed.

lukpueh commented:
Just merged #1708. Please rebase and mark ready for review!

MVrachev commented:
> Thanks for the PR, @MVrachev! Looks like a straight-forward set of changes. I left some minor suggestions, but will approve regardless.
> 
> I see you adopted the filename changes (add "_old" infix) in some doc headers but not all. I suggest to do this consistently and I'm fine with not updating the doc headers, given that these files are bound to be removed.

I rebased on top of develop and I think changed all the relevant doc headers and places where the file names of
the old test files were mentioned.
The pr is ready for review!

MVrachev commented:
I had to rebase to change the test name used as an example in `docs/CONTRIBUTIONS.rst`.

Commit:
Add logging configuration in tests/test_examples

Logging configuration is part from each test file begging from commit
https://github.com/theupdateframework/python-tuf/commit/03b15fb4be88fa2273134d23025c26d3bb2d8f8a
We should that as well for tests/test_examples.py

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented on tests/test_examples.py:
```diff
@@ -7,11 +7,14 @@
 
+from tests import utils
+
 
```
I linted with `isort` this file and that was what suggested.
I thought the list should be like that:
```
import glob
import os
import shutil
import sys
import tempfile
import unittest

from pathlib import Path
from typing import ClassVar, List
from tests import utils
```
but isort fixes that to the current state.

MVrachev commented on tests/test_examples.py:
```diff
@@ -7,11 +7,14 @@
 from typing import ClassVar, List
 
+from tests import utils
+
```
I linted with `isort` this file and that was what suggested.
I thought the list should be like that:
```
import glob
import os
import shutil
import sys
import tempfile
import unittest

from pathlib import Path
from typing import ClassVar, List
from tests import utils
```
but isort fixes that to the current state.

jku commented on pyproject.toml:
```diff
@@ -88,6 +88,7 @@ exclude=".*_old.py"
   "securesystemslib.*",
-  "urllib3.*"
+  "urllib3.*",
+  "dateutil.relativedelta",
```
```
tests/test_api.py:19: error: Library stubs not installed for "dateutil.relativedelta" (or incompatible with Python 3.10)  [import]
tests/test_api.py:19: note: Hint: "python3 -m pip install types-python-dateutil"
```

I would try adding types-python-dateutil into requirements-test.txt instead

MVrachev replied:
Great, I totally forgot I can do this.

jku commented on tests/test_api.py:
```diff
@@ -136,7 +136,12 @@ def test_compact_json(self) -> None:
 
     def test_read_write_read_compare(self) -> None:
-        for metadata in [Root.type, Snapshot.type, Timestamp.type, Targets.type]:
+        for metadata in [
+            Root.type,
+            Snapshot.type,
+            Timestamp.type,
+            Targets.type,
+        ]:
```
it's a nitpick but in cases like this and the one below a bit of tweaking could save ten lines of code:

This probably fits and is a better name:
`        for role in [Root.type, Snapshot.type, Timestamp.type, Targets.type]:`
or :
`        for rolename in TOP_LEVEL_ROLE_NAMES:`

Maybe this one is not worth modifying the PR for but this sort of thing adds up... and makes the tests harder to read

MVrachev replied:
`for role in [Root.type, Snapshot.type, Timestamp.type, Targets.type]:` is more than 80 characters, but
`for rolename in TOP_LEVEL_ROLE_NAMES:` will do the work.
I thought we have deleted `TOP_LEVEL_ROLE_NAMES` when we introduce the constatnts and it's a nice surprise we didn't.

jku reviewed (changes requested):
Moving almost obsolete files is a bit annoying but I can see why you want to flick the switch on this one :) style guidelines won't be observed without hard failures... LGTM in general

We should keep statically checking dateutil, otherwise my comments are just discussion.

Commit:
Simulator: make _fetch_{metadata,target} public

Make _fetch_metadata and _fetch_taget public by renaming them to
fetch_metadata and fetch_target.
This will allow the removal of multiple pylint disables because of
"accessing private members".

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Mypy: disable attr-defined warning

Mypy warns us when we assign a not defined variable to an object, but
that is something that we are warned for from pylint (seach for
"pylint: disable=no-member" in test_updater_key_rotations.py
and you will find an example where we have to disable it).
We don't want to have two linters checking for the same thing
as we can end up disabling two warnings that are actually the same
on a single line.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Metadata API: change bump_expiration delta type

Change the delta argument type from the tuf/api/metadata.py module
in Signed.bump_expiration() to include relativedelta as
this provides an easier interface for the callers.
We are already testing for that inside test/api line 338.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have rebased on top of develop.
It was painful to do so as there were multiple changes and there were new warnings from all of the linters meaning I had to update their commits.
Additionally, I addressed all of Jussi's comments and had to add a couple of new commits (the last two) that handle specific problems.
It would be great if we merge this soon as with each new change it's annoying to do that work again.
@jku and/or @lukpueh can you have a new look?

lukpueh commented on tests/test_updater_old.py:
```diff
@@ -21,7 +21,7 @@
 
 <Purpose>
-  'test_updater.py' provides a collection of methods that test the public /
+  'test_updater.py_old' provides a collection of methods that test the public /
```
Is this the coding equivalent of the french Verlan? :P Given the short remaining life time of this doc, I say we leave this as is.

lukpueh reviewed (approved):
Many thanks for doing this tedious work, Martin! I'll go forward and merge, because it looks like you addressed Jussi's requested change.
